---
title: "Get TF-motif annotation"
output: 
  html_document:
    toc: true
    toc_float: true
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE)

library(data.table)
library(dplyr)
library(ggpubr)
library(TF2gene)
```


## Motifs with PWM

```{r}
meta_motif = get_cb_motif_info('/fh/fast/sun_w/kenny_zhang/v10nr_clust_public/singletons', ncores = 3)

# usethis::use_data(meta_motif, overwrite = T)

meta_motif
```


## Motifs with TF annotation

```{r}
tf2motif = fread('/fh/fast/sun_w/kenny_zhang/motifs-v10nr_clust-nr.hgnc-m0.001-o0.0.tbl') %>%
  as_tibble() %>%
  rename(motif_id = '#motif_id')

tf2motif

# Number of TFs
length(unique(tf2motif$gene_name))

# Number of motifs
length(unique(tf2motif$motif_id))
```


## Subset motifs

```{r}
tf2motif = tf2motif %>%
  filter(motif_id %in% meta_motif$motif)

# Number of TFs
length(unique(tf2motif$gene_name))

# Number of motifs
length(unique(tf2motif$motif_id))
```


## Split TF-motif annotation by type

```{r}
types = c('gene is directly annotated', 
          'motif similar to', 
          'gene is orthologous to', 
          'motif is annotated for orthologous gene',
          'gene is annotated for similar motif')

num_types = sapply(types, function(type) sum(startsWith(tf2motif$description, type)))

sum(num_types) == nrow(tf2motif)

sort(num_types, decreasing = T)

tf2motif$type = sapply(tf2motif$description, function(description) {
  which_type = which(sapply(types, function(type) startsWith(description, type)))
  types[which_type]
})

# df_tf2motif = tf2motf
# usethis::use_data(df_tf2motif, overwrite = T)
```


## TF degree

```{r}
# Overall
plot_data = tf2motif %>% 
  select(motif_id, gene_name) %>%
  unique() %>%
  group_by(gene_name) %>%
  summarise(tf_degree = n(), .groups = 'drop')
  
ggplot(plot_data, aes(tf_degree)) +
  geom_histogram(bins = 100) +
  labs(x = 'TF degree') +
  theme_pubr()
```

```{r, fig.width=10, fig.height=6}
# Each type
plot_data = tf2motif %>% 
  select(motif_id, gene_name, type) %>%
  unique() %>%
  group_by(type, gene_name) %>%
  summarise(tf_degree = n(), .groups = 'drop')
  
ggplot(plot_data, aes(tf_degree)) +
  geom_histogram(bins = 100) +
  facet_wrap(~type, nrow = 2, ncol = 3, scales = 'free_x') +
  labs(x = 'TF degree') +
  theme_pubr()
```


## Motif degree

```{r}
# Overall
plot_data = tf2motif %>% 
  select(motif_id, gene_name) %>%
  unique() %>%
  group_by(motif_id) %>%
  summarise(motif_degree = n(), .groups = 'drop')
  
ggplot(plot_data, aes(motif_degree)) +
  geom_histogram(bins = 100) +
  labs(x = 'Motif degree') +
  theme_pubr()
```

```{r, fig.width=10, fig.height=6}
# Each type
plot_data = tf2motif %>% 
  select(motif_id, gene_name, type) %>%
  unique() %>%
  group_by(type, motif_id) %>%
  summarise(motif_degree = n(), .groups = 'drop')
  
ggplot(plot_data, aes(motif_degree)) +
  geom_histogram(bins = 100) +
  facet_wrap(~type, nrow = 2, ncol = 3, scales = 'free_x') +
  labs(x = 'Motif degree') +
  theme_pubr()
```


## Calculate TF2motif matrix

```{r}
tfs = sort(unique(tf2motif$gene_name))
motifs = sort(unique(tf2motif$motif_id))

# By direct annotation
df_direct = tf2motif %>% 
  filter(type == 'gene is directly annotated') %>%
  select(gene_name, motif_id) %>%
  unique()

mat_direct = Matrix::sparseMatrix(
  i = match(df_direct$gene_name, tfs),
  j = match(df_direct$motif_id, motifs),
  dims = c(length(tfs), length(motifs)),
  dimnames = list(tfs, motifs)
)

# By orthology
df_orth = tf2motif %>% 
  filter(type %in% c('gene is orthologous to', 'motif is annotated for orthologous gene')) %>%
  select(gene_name, motif_id) %>%
  unique()

mat_orth = Matrix::sparseMatrix(
  i = match(df_orth$gene_name, tfs),
  j = match(df_orth$motif_id, motifs),
  dims = c(length(tfs), length(motifs)),
  dimnames = list(tfs, motifs)
)

# By similarity
df_simi = tf2motif %>% 
  filter(type %in% c('motif similar to', 'gene is annotated for similar motif')) %>%
  select(gene_name, motif_id) %>%
  unique()

mat_simi = Matrix::sparseMatrix(
  i = match(df_simi$gene_name, tfs),
  j = match(df_simi$motif_id, motifs),
  dims = c(length(tfs), length(motifs)),
  dimnames = list(tfs, motifs)
)

# Combine
mat = pmax(3 * mat_direct, 2 * mat_orth, mat_simi)

# mat_tf2motif = mat
# usethis::use_data(mat_tf2motif, overwrite = T)
```
